packages/flutter/lib/src/material/ink_sparkle.dart

```diff
@@ -0,0 +1,3673 @@
+      uSparkleAlpha: _sparkleAlpha.value,
+
+      // The following uniforms are driven by the turbulence phase and change
+      // each frame of the animation. Theese unifroms are uses to modify the
+      // default (if these fields are unset or 0) circular outer ring to a
+      // non-uniform shape that is more like an actual ink splash. In addition
+      // to the positional based triangular noise created in the shader, these
+      // uniforms also vary the appearence of the sparkles even when the same
+      // location is tapped.
+      uTurbulencePhase: turbulencePhase,
+      uNoisePhase: noisePhase / 1000.0,
+      uCircle1: Vector2(
+        turbulenceScale * 0.5 + (turbulencePhase * 0.01 * math.cos(turbulenceScale * 0.55)),
+        turbulenceScale * 0.5 + (turbulencePhase * 0.01 * math.sin(turbulenceScale * 0.55)),
+      ),
+      uCircle2: Vector2(
+        turbulenceScale * 0.2 + (turbulencePhase * -0.0066 * math.cos(turbulenceScale * 0.45)),
+        turbulenceScale * 0.2 + (turbulencePhase * -0.0066 * math.sin(turbulenceScale * 0.45)),
+      ),
+      uCircle3: Vector2(
+        turbulenceScale + (turbulencePhase * -0.0066 * math.cos(turbulenceScale * 0.35)),
+        turbulenceScale + (turbulencePhase * -0.0066 * math.sin(turbulenceScale * 0.35)),
+      ),
+      uRotation1: Vector2(math.cos(rotation1), math.sin(rotation1)),
+      uRotation2: Vector2(math.cos(rotation2), math.sin(rotation2)),
+      uRotation3: Vector2(math.cos(rotation3), math.sin(rotation3)),
+    );
+  }
+
+  Vector4 _colorToVector4(Color color) {
+    return Vector4(
+        color.red / 255.0,
+        color.blue / 255.0,
+        color.green / 255.0,
+        color.alpha / 255.0,
+      );
+  }
+
+  /// Transforms the canvas for an ink feature to be painted on the [canvas].
+  ///
+  /// This should be called before painting ink features that do not use
+  /// [paintInkCircle].
+  ///
+  /// The [transform] argument is the [Matrix4] transform that typically
+  /// shifts the coordinate space of the canvas to the space in which
+  /// the ink feature is to be painted.
+  ///
+  /// For examples on how the function is used, see [InkSparkle] and [paintInkCircle].
+  void _transformCanvas({
+    required Canvas canvas,
+    required Matrix4 transform,
+  }) {
+    final Offset? originOffset = MatrixUtils.getAsTranslation(transform);
+    if (originOffset == null) {
+      canvas.transform(transform.storage);
+    } else {
+      canvas.translate(originOffset.dx, originOffset.dy);
+    }
+  }
+
+
+  /// Clips the canvas for an ink feature to be painted on the [canvas].
+  ///
+  /// This should be called before painting ink features with [paintFeature]
+  /// that do not use [paintInkCircle].
+  ///
+  /// [clipCallback] is the callback used to obtain the [Rect] used for clipping the ink effect.
+  /// If [clipCallback] is null, no clipping is performed on the ink circle.
+  ///
+  /// Clipping can happen in 3 different ways:
```

Comment:
nit: macro this to avoid having to sync documentation with `InkSparkle`

Reply:
I think it makes sense to only keep this note in InkSparkle now, and then it will possibly be in one place or by way of a macro as part of https://github.com/flutter/flutter/issues/73163